pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+		attributes); err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else if decision != authorizer.DecisionAllow {
+		responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+		return
+	}
+
+	if retryAfterSeconds, err := h.tryToCreate(ctx, user.GetName(), lcluster.Name, workspaceType); err != nil {
+		responsewriters.ErrorNegotiated(err, errorCodecs, schema.GroupVersion{}, rw, req)
+		return
+	} else {
+		// Return a 429 status asking the client to try again after the creationDelay
+		rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+		http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+		return
+	}
+}
+
+// reHomeWorkspaceNameDisallowedtChars is the regexp that defines what characters
+// are disallowed in a home workspace name.
+// Home workspace name is derived from the user name, with disallowed characters
+// replaced.
+var reHomeWorkspaceNameDisallowedtChars = regexp.MustCompile("[^a-z0-9-]")
+
+// getHomeLogicalClusterName returns the logicalcluster name of the home workspace for a given user
+// The home workspace logical cluster ancestors are home bucket workspaces whose name is based
+// on the user name sha1 hash.
+func (h *homeWorkspaceHandler) getHomeLogicalClusterName(userName string) logicalcluster.Name {
+	// bucketLevels <= 5
+	// bucketSize <= 4
+	bytes := sha1.Sum([]byte(userName))
+
+	result := h.homePrefix
+	for level := 0; level < h.bucketLevels; level++ {
+		var bucketBytes = make([]byte, h.bucketSize)
+		bucketBytesStart := level
+		bucketCharInteger := binary.BigEndian.Uint32(bytes[bucketBytesStart : bucketBytesStart+4])
+		for bucketCharIndex := 0; bucketCharIndex < h.bucketSize; bucketCharIndex++ {
+			bucketChar := byte('a') + byte(bucketCharInteger%26)
+			bucketBytes[bucketCharIndex] = bucketChar
+			bucketCharInteger /= 26
+		}
+		result = result.Join(string(bucketBytes))
+	}
+
+	userName = reHomeWorkspaceNameDisallowedtChars.ReplaceAllLiteralString(userName, "-")
+	userName = strings.TrimLeftFunc(userName, func(r rune) bool {
+		return r <= '9'
+	})
+	userName = strings.TrimRightFunc(userName, func(r rune) bool {
+		return r == '-'
+	})
+
+	return result.Join(userName)
+}
+
+// needsAutomaticCreation deduces, from the logical cluster name,
+// according the the expected home root and home bucket level number,
+// whether the corresponding workspace has to be checked for automatic creation
+// and what its workspace type will be.
+func (h *homeWorkspaceHandler) needsAutomaticCreation(logicalClusterName logicalcluster.Name) (needsCheck bool, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) {
+	if !logicalClusterName.HasPrefix(h.homePrefix) {
+		return false, ""
+	}
+
+	if logicalClusterName == h.homePrefix {
+		return false, ""
+	}
+
+	levelsToHomePrefix := 0
+	for lcluster := logicalClusterName; lcluster != h.homePrefix; lcluster, _ = lcluster.Split() {
```

Comment:
a) couldn't this just count the number of `:` in the name?
b) seems like a useful helper for the `logicalcluster` library

Reply:
have only seen this occurence of that need. If we find 3 others let's talk about `logicalcluster.Name.Levels() int`